lib/libvarnishapi/vsl_dispatch.c

```diff
@@ -1274,15 +1274,54 @@ vslq_shmref_check(struct VSLQ *vslq)
 	return (vsl_more);
 }
 
+static unsigned
+vslq_candidate(struct VSLQ *vslq, const uint32_t *ptr)
+{
+	enum VSL_transaction_e type;
+	enum VSL_reason_e reason;
+	struct VSL_data *vsl;
+	enum VSL_tag_e tag;
+	unsigned p_vxid;
+	int i;
+
+	CHECK_OBJ_NOTNULL(vslq, VSLQ_MAGIC);
+	AN(ptr);
+
+	assert(vslq->grouping != VSL_g_raw);
+	if (vslq->grouping != VSL_g_vxid)
+		return (1);
+
+	vsl = vslq->vsl;
+	CHECK_OBJ_NOTNULL(vsl, VSL_MAGIC);
+	if (!vsl->c_opt && !vsl->b_opt)
```

Comment:
Isn't there an argument that if (grouping != VSL_g_sess && type == VSL_t_sess), then we should not keep it? Regardless of the state of c_opt and b_opt. It looks to me like we may have been doing a lot of unnecessary session type tracking as was before, and also with this patch.

Reply:
Before, we definitely tracked sessions for nothing in this scenario. However this patch should prevent session tracking when we don't need them. At least according to my testing.

I focused on vxid grouping because that's where `varnishncsa` operates by default, but it might still be the case for request grouping that sessions are collected for no good reason.

Reply:
I tested this, and confirmed that we do still with this patch set, keep track of session transactions for request grouping. As these will never be reported (filtered out in vslq_callback()), that work is wasted.

I wonder if this function should look something like:

```
static unsigned
vslq_candidate(struct VSLQ *vslq, const uint32_t *ptr)
{
	enum VSL_transaction_e type;
	enum VSL_reason_e reason;
	struct VSL_data *vsl;
	enum VSL_tag_e tag;
	unsigned p_vxid;
	int i;

	CHECK_OBJ_NOTNULL(vslq, VSLQ_MAGIC);
	AN(ptr);

	assert(vslq->grouping != VSL_g_raw);
	if (vslq->grouping == VSL_g_session)
		/* All are needed */
		return (1);

	vsl = vslq->vsl;
	CHECK_OBJ_NOTNULL(vsl, VSL_MAGIC);
	if (vslq->grouping == VSL_g_vxid) {
		if (!vsl->c_opt && !vsl->b_opt)
			/* Implies also !vsl->E_opt */
			return (1);
		if (vsl->c_opt && !VSL_CLIENT(ptr))
			return (0);
		if (vsl->b_opt && !VSL_BACKEND(ptr))
			return (0);
		/* Need to parse the Begin tag - fallthrough to below */
	}

	tag = VSL_TAG(ptr);
	assert(tag == SLT_Begin);
	i = vtx_parse_link(VSL_CDATA(ptr), &type, &p_vxid, &reason);
	if (i != 3 || type == VSL_t_unknown)
		return (0);

	if (type == VSL_t_sess)
		return (0);

	if (vslq->grouping == VSL_g_vxid && reason == VSL_r_esi && !vsl->E_opt)
		return (0);

	return (1);
}
```

Reply:
We had a discussion about this at some point, I thought you could make `-q`uery sessions but not display them with `-g request` and @mbgrydeland told me irrelevant transactions are removed before testing queries.

So yes, I agree with the suggestion.

Reply:
I pushed 55494ae1a1f5184afc2fdc831241e2b912d07a4b on @mbgrydeland's behalf with the following change:

```diff
        if (vslq->grouping == VSL_g_vxid) {
                if (!vsl->c_opt && !vsl->b_opt)
                        return (1); /* Implies also !vsl->E_opt */
-               if (vsl->c_opt && !VSL_CLIENT(ptr))
+               if (!vsl->b_opt && !VSL_CLIENT(ptr))
                        return (0);
-               if (vsl->b_opt && !VSL_BACKEND(ptr))
+               if (!vsl->c_opt && !VSL_BACKEND(ptr))
                        return (0);
                /* Need to parse the Begin tag - fallthrough to below */
        }
```

Otherwise candidates are discarded too early. This is essentially what my code would do, shaped slightly differently.